chore: development SDK generation script patch

GBirkel opened the pull request:
The front end repo contains a script: `./scripts/generate-nestjs-sdk.bash`  It runs the OpenAPI API Generator inside a docker container.  It's for when you're making API changes in the back end and want to develop with those changes in the front end.

The generator assumes you're already running a build of the back end on your local machine, and calls it, asking for the API spec.  Then the script builds the result, and edits your local node module cache, replacing the "official" `@scicatproject/scicat-sdk-ts-angular` with the substitute version.

It's weird, but it's convenient: You don't need to upload a package to a repository.

Unfortunately, it didn't work with SciCat Live, because there's no way to map the "backend.localhost" name generated by the proxy into the docker container. So I modified the script to fetch the spec with a curl command into a file, and give the file to the OpenAPI Generator instead of giving it a URL. Easy fix.  Since SciCatLive runs the back end at a different address, I added a command-line flag for invoking the script:  `--scicatlive`.

GBirkel committed:
Development SDK generation script patch, for playing nice with SciCatLive.

Junjiequan commented:
I prefer not to include any service-specific dependent code in the Scicat core, and instead provide a more generic solution. In this case, allowing users to provide arbitrary URLs might be better, as it would help avoid future requests to add new URLs

GBirkel commented:
> I prefer not to include any service-specific dependent code in the Scicat core, and instead provide a more generic solution. In this case, allowing users to provide arbitrary URLs might be better, as it would help avoid future requests to add new URLs

I agree that this would be a cleaner implementation, but on the other hand, we're talking about a developers-only support script.  It should work out-of-the-box as much as possible.

I could modify it to accept an arbitrary URL, but that would also mean adding documentation to SciCatLive somewhere explaining that a specific, always the same, URL should be passed to this script on the command line when invoking.  Why make developers go to the trouble?  Are we supporting SciCatLive (one of our own projects), or are we not?

GBirkel committed:
Forgot to include a few lines from my development branch...

GBirkel committed:
Merge branch 'master' into fix-dev-sdk-for-scicatlive

nitrosx commented on scripts/generate-nestjs-sdk.bash:
Changes to this file are also included in this PR https://github.com/SciCatProject/frontend/pull/1910/files#top.
I left a comment there also, but just in case I leave them here also.

My apologies for being direct, but may ask you to submit cleaner PRs and make sure that changes are not duplicated across PRs?

GBirkel replied:
Yeah, that snuck in there because I needed that code on my branch in order to develop with SciCatLive.  It shouldn't have been committed.  Once it's more complete, I can re-make the PR if necessary.

nitrosx commented on scripts/generate-nestjs-sdk.bash:
```diff
@@ -2,12 +2,38 @@
 #
 #
 
+USING_SCICAT_LIVE=0
+while [[ "$#" -gt 0 ]]; do
+	case "$1" in
+		--scicatlive)
+			USING_SCICAT_LIVE=1
+			shift
+			;;
+		*)
+			echo "Unknown argument: $1"
+			exit 1
+			;;
+	esac
+done
+
```
I would change this section of the code like below. This way we have a default, but we do not hardcode the all the other possible values, but we provide the user the option to specify the correct URL, including the case where the backend is running on a different port or different url.

```
SWAGGER_API_URL="http://localhost:3000/explorer-json"
while [[ "$#" -gt 0 ]]; do
	case "$1" in
		--swagger-url)
      shift
      SWAGGER_API_URL=$1
			shift
			;;
		*)
			echo "Unknown argument: $1"
			exit 1
			;;
	esac
done
```

nitrosx commented on scripts/generate-nestjs-sdk.bash:
```diff
@@ -2,12 +2,38 @@
+
+echo -e "\nFetching the API from local instance of back end..."
+
+if [ $USING_SCICAT_LIVE -eq 1 ];
+then
+	# For when developing with SciCat Live:
+	curl http://backend.localhost/explorer-json > local-api-for-generator.json
+else
+	# For when developing with a backend running directly on localhost:
+	curl http://host.docker.internal:3000/explorer-json > local-api-for-generator.json
+fi
```
change the code to:
```
curl ${SWAGGER_API_URL} > local-api-for-generator.json
```

nitrosx reviewed (dismissed):
Once the requested changes are addressed, I will approve it.
Thanks for helping out

nitrosx commented:
> I agree that this would be a cleaner implementation, but on the other hand, we're talking about a developers-only support script. It should work out-of-the-box as much as possible.
> 
> I could modify it to accept an arbitrary URL, but that would also mean adding documentation to SciCatLive somewhere explaining that a specific, always the same, URL should be passed to this script on the command line when invoking. Why make developers go to the trouble? Are we supporting SciCatLive (one of our own projects), or are we not?

I would much prefer a cleaner solution. The change that you propose (and elaborated further in my comments) is the most sensible.

Regarding SciCatLive, it is an official SciCat supported project, but it is not part of core... at least not yet.
I would argue that any scripts in core (FE or BE) should be completely agnostic to SciCatLive, although I agree with you regarding passing the same URL to the script when working in SciCatLive.

I suggest to change the script here to accept an arbitrary URL and add a dedicated script in SciCatLive to call this script with the correct URL specific to SciCatLive.
That should implement a solution addressing all concerns.

GBirkel commented:
> I suggest to change the script here to accept an arbitrary URL and add a dedicated script in SciCatLive to call this script with the correct URL specific to SciCatLive. That should implement a solution addressing all concerns.

That works for me!  I'll change this code and create a ticket for SciCatLive, and assign it to myself.

GBirkel committed:
Accepting an alternate Swagger URL via command line argument.

minottic commented on scripts/generate-nestjs-sdk.bash:
```diff
@@ -17,14 +36,25 @@ echo -e "\nGenerating the new sdk..."
 	-v "`pwd`:/local" \
 	openapitools/openapi-generator-cli:v7.13.0 generate \
-	-i http://host.docker.internal:3000/explorer-json \
+	-i /local/local-api-for-generator.json \
```
sorry, there was a typo in my previous message. Try with 
`BACKEND_DOCKER_URL: http://backend:3000`
This is the URL where the backend is exposed in the docker network, so generate should be able to call it from within the container

GBirkel replied:
That does not work.  Docker does not give this separately instantiated standalone container access to the same network that SciCatLive was created with when it was instantiated.

minottic replied:
Where are you running this command from? Are you running it from within the frontend container created by scicatlive in DEV mode? That should work...

minottic replied:
I have played around a bit with the volume problem, and I can suggest a solution if you don't have already, which we can discuss. About the HOST part, I am honestly fine with both solutions, either storing it into a file or setting the add-host var dynamically. I would suggest in both ways to enable setting it with an ENV var, so ti will become transparent for the user in scicatlive

minottic replied:
if you run scicatlive with DEV=true, all scicat related containers can be used to develop into (we can follow this up on slack, probably the documentation is not super clear). So with DEV=true, you could attach in the frontend container and develop your changes there. If you then need a new build of the SDK, you could call `npm run generate:sdk:local` from within the scicatlive frontend container and this would mount the SDK in the correct place. [This](https://github.com/SciCatProject/frontend/compare/master...sdk_generate_host) is the change that the frontend would require (or yours with a small mod) and [this](https://github.com/SciCatProject/scicatlive/compare/main...generate_sdk) the scicatlive one. In this way we could support scicatlive sdk rebuild and set the env var inplace of the user

GBirkel replied:
I'm afraid the solution has drifted too much for me to comment, because I don't know enough about the "dev" mode in SciCatLive.  Months ago when I first tried to parse through what it was doing I got confused, which is why I set up https://github.com/als-computing/scicatlive-als .  All it does is map two local repos - front and back end - into a Docker environment, which is enough to run while developing.

The intention behind the PR I submitted was to make the API generation script work with an instance of the back end running in a docker container, which it does by using an intermediate file.  If this PR is modified to remove the intermediate file, my development scenario will remain broken, but in the end that doesn't really matter I suppose because I can just patch the script on my own branch in the als-computing fork.

minottic replied:
what I meant was that this PR is fine for me. I was only suggesting enabling also setting the SWAGGER from an ENV so I could later reuse in scicatlive. But no need for it to be in this PR, I will open one later on top of your changes to implement the cascading mechanism with the ENV var so we can support it in scicatlive. 

I wonder how you keep your fork in sync, maybe you could propose your changes directly in scicatlive (the local repos mount) and we see how to make them fit?

GBirkel replied:
When I pull changes into the ALS-Customizations branch the changes are almost always to files that I've removed.  I do usually have to manually merge the remaining changes, but it doesn't take very long.  So I'm ok with it. :)

I don't know if it would be worth merging it back into the main repo, honestly?  It's convenient for me, developing on MacOS, but it's pointless for Windows people for example...

GBirkel committed:
Version bump for OpenAPI generator.

GBirkel committed:
Useful message when build output is not found.  Halts before deleting, so user can run attempt manual build.

GBirkel committed:
A debugging note.

GBirkel committed:
Merge branch 'master' into fix-dev-sdk-for-scicatlive

GBirkel committed:
Further refinements to the SDK generation script.

GBirkel committed:
Minor fix to SDK copy paths.

GBirkel committed:
Merge branch 'master' into fix-dev-sdk-for-scicatlive

bpedersen2 commented on scripts/generate-nestjs-sdk.bash:
```diff
@@ -1,13 +1,39 @@
+fi
+
+# This default is for when developing with a backend running directly on localhost
+SWAGGER_API_URL="http://localhost:3000/explorer-json"
```
Consider 
```
SWAGGER_API_URL=${SWAGGER_API_URL:-"http://localhost:3000/explorer-json"} 

```
Then one can override it with a environment variable as well.

GBirkel committed:
Allowing SWAGGER_API_URL environment variable.

GBirkel committed:
Using the same config file that the back end CI code uses to configure the API generator.

GBirkel committed:
Merge branch 'master' into fix-dev-sdk-for-scicatlive

nitrosx commented:
Closing this PR due to inactivity.
Keen to re-open it if author or other contributors will work on.
